replay/models/cql.py

```diff
@@ -0,0 +1,357 @@
+
+        # predict items relevance for the user
+        with torch.no_grad():
+            user_item_pairs['relevance'] = model.forward(items_batch).numpy()
+
+        return user_item_pairs
+
+    # pylint: disable=too-many-arguments
+    def _predict(
+        self,
+        log: DataFrame,
+        k: int,
+        users: DataFrame,
+        items: DataFrame,
+        user_features: Optional[DataFrame] = None,
+        item_features: Optional[DataFrame] = None,
+        filter_seen_items: bool = True,
+    ) -> DataFrame:
+        available_items = items.toPandas()["item_idx"].values
+        policy_bytes = self.serialize_policy()
+
+        def grouped_map(log_slice: pd.DataFrame) -> pd.DataFrame:
+            return CQL._predict_pairs_inner(
+                model=policy_bytes,
+                user_idx=log_slice["user_idx"][0],
+                items=available_items,
+            )[["user_idx", "item_idx", "relevance"]]
+
+        # predict relevance for all available items and return them as is;
+        # `filter_seen_items` and top `k` params are ignored
+        self.logger.debug("Predict started")
+        return users.groupby("user_idx").applyInPandas(grouped_map, REC_SCHEMA)
+
+    def _predict_pairs(
+        self,
+        pairs: DataFrame,
+        log: Optional[DataFrame] = None,
+        user_features: Optional[DataFrame] = None,
+        item_features: Optional[DataFrame] = None,
+    ) -> DataFrame:
+        policy_bytes = self.serialize_policy()
+
+        def grouped_map(user_log: pd.DataFrame) -> pd.DataFrame:
+            return CQL._predict_pairs_inner(
+                model=policy_bytes,
+                user_idx=user_log["user_idx"][0],
+                items=np.array(user_log["item_idx_to_pred"][0]),
+            )[["user_idx", "item_idx", "relevance"]]
+
+        self.logger.debug("Calculate relevance for user-item pairs")
+        return (
+            pairs
+            .groupBy("user_idx")
+            .agg(sf.collect_list("item_idx").alias("item_idx_to_pred"))
+            .join(log.select("user_idx").distinct(), on="user_idx", how="inner")
+            .groupby("user_idx")
+            .applyInPandas(grouped_map, REC_SCHEMA)
+        )
+    
+    @property
+    def _init_args(self):
+        args = dict(
+            top_k=self.top_k,
+            action_randomization_scale=self.action_randomization_scale,
+            n_epochs=self.n_epochs,
+        )
+        args.update(**self.model.get_params())
+        return args
+
+    def serialize_policy(self) -> bytes:
+        # store using temporary file and immediately read serialized version
+        with tempfile.NamedTemporaryFile(suffix='.pt') as tmp:
+            # noinspection PyProtectedMember
+            self.model._impl.save_policy(tmp.name)
+            with open(tmp.name, 'rb') as policy_file:
+                return policy_file.read()
+
+    @staticmethod
+    def assert_omp_single_thread():
```

Comment:
i think you this method hasn't to be here. it is common method  for all models with pytorch. Please add it to utils

Reply:
Done!